UnitySDK/Assets/ML-Agents/Scripts/InferenceBrain/BarracudaModelParamLoader.cs

```diff
@@ -0,0 +1,615 @@
+            }
+            else
+            {
+                tensorTester[TensorNames.ActionOutput] = CheckDiscreteActionOutputShape;
+            }
+            // If the model expects an output but it is not in this list
+            foreach (var name in _model.outputs)
+            {
+                if (tensorTester.ContainsKey(name))
+                {
+                    var tester = tensorTester[name];
+                    var error = tester.Invoke(_model.GetShapeByName(name), modelActionSize);
+                    if (error != null)
+                    {
+                        _failedModelChecks.Add(error);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Checks that the shape of the discrete action output is the same in the
+        /// model and in the Brain Parameters.
+        /// </summary>
+        /// <param name="shape"> The tensor shape that is expected by the model</param>
+        /// <param name="modelActionSize"> The size of the action output that is expected
+        /// by the model.</param>
+        /// <returns>If the Check failed, returns a string containing information about why the
+        /// check failed. If the check passed, returns null.</returns>
+        private string CheckDiscreteActionOutputShape(TensorShape shape, int modelActionSize)
+        {
+            var bpActionSize = _brainParameters.vectorActionSize.Sum();
+            if  (modelActionSize != bpActionSize)
+            {
+                return string.Format(
+                    "Action Size of the model does not match. " +
+                    "The BrainParameters expect {0} but the model contains {1}.",
+                    bpActionSize, modelActionSize);
+            }
+            return null;
+        }
+        
+        /// <summary>
+        /// Checks that the shape of the continuous action output is the same in the
+        /// model and in the Brain Parameters.
+        /// </summary>
+        /// <param name="shape"> The tensor shape that is expected by the model</param>
+        /// <param name="modelActionSize"> The size of the action output that is expected
+        /// by the model.</param>
+        /// <returns>If the Check failed, returns a string containing information about why the
+        /// check failed. If the check passed, returns null.</returns>
+        private string CheckContinuousActionOutputShape(TensorShape shape, int modelActionSize)
+        {
+            var bpActionSize = _brainParameters.vectorActionSize[0];
+            if  (modelActionSize != bpActionSize)
+            {
+                return string.Format(
+                    "Action Size of the model does not match. " +
+                    "The BrainParameters expect {0} but the model contains {1}.",
+                    bpActionSize, modelActionSize);
+            }
+            return null;
+        }
+    }
+}
+
+public class BarracudaUtils
```

Comment:
This should be in its own script

Reply:
I expect these to live for one release only, after we move to Barracuda Tensor across the board most of this code will disappear..